File: data structures and algorithms/cpp/06_018.cpp
#include <vector>
#include <stack>

/**
 * A basic binary tree node implementation.
 */
class TreeNode {
public:
   int value;
   TreeNode* left;
   TreeNode* right;
   
   // Constructor
   TreeNode(int value) {
      this->value = value;
      this->left = nullptr;
      this->right = nullptr;
   }
};

/**
 * Implementation of Depth-First Search (DFS) tree 
 * traversal.
 */
class DFSTree {
public:
   /**
    * Perform recursive DFS traversal in pre-order.
    */
   std::vector<int> dfsRecursive(TreeNode* root) {
      std::vector<int> result;
      explore(root, result);
      return result;
   }
   
private:
   void explore(TreeNode* node, std::vector<int>& result) {
      if (node == nullptr) {
         return;
      }
      
      // Pre-order: Process BEFORE recursion
      result.push_back(node->value);
      
      // Go as deep as possible left
      explore(node->left, result);
      // Then go right
      explore(node->right, result);
   }
   
public:
   /**
    * Perform iterative DFS traversal in pre-order
    * using a stack.
    */
   std::vector<int> dfsIterative(TreeNode* root) {
      if (root == nullptr) {
         return std::vector<int>();
      }
      
      std::vector<int> result;
      std::stack<TreeNode*> stack;
      // Use stack to mimic recursion
      stack.push(root);
      
      while (!stack.empty()) {
         // Get last node added
         TreeNode* node = stack.top();
         stack.pop();
         result.push_back(node->value);
         
         // Add right first (so left gets processed first)
         if (node->right != nullptr) {
            stack.push(node->right);
         }
         if (node->left != nullptr) {
            stack.push(node->left);
         }
      }
      
      return result;
   }
};

File: data structures and algorithms/cs/06_018.cs
using System;
using System.Collections.Generic;

public class DFSTree
{
   public List<int> DfsRecursive(TreeNode root)
   {
      var result = new List<int>();
      Explore(root, result);
      return result;
   }

   private void Explore(TreeNode node, List<int> result)
   {
      if (node == null)
      {
         return;
      }

      // Pre-order: Process node first
      result.Add(node.Value);

      // Then go left and right
      Explore(node.Left, result);
      Explore(node.Right, result);
   }

   public List<int> DfsIterative(TreeNode root)
   {
      if (root == null)
      {
         return new List<int>();
      }

      var result = new List<int>();
      var stack = new Stack<TreeNode>();
      stack.Push(root);

      while (stack.Count > 0)
      {
         var node = stack.Pop();
         result.Add(node.Value);

         if (node.Right != null)
         {
            stack.Push(node.Right);
         }
         if (node.Left != null)
         {
            stack.Push(node.Left);
         }
      }

      return result;
   }
}

public class TreeNode
{
   public int Value { get; set; }
   public TreeNode Left { get; set; }
   public TreeNode Right { get; set; }

   public TreeNode(int value)
   {
      Value = value;
      Left = null;
      Right = null;
   }
}
